native/src/main/java/io/ballerina/stdlib/http/transport/contractimpl/sender/states/ReceivingEntityBody.java

```diff
@@ -105,4 +127,93 @@ public void handleIdleTimeoutConnectionClosure(TargetHandler targetHandler,
+            outboundAccessLogMessage.setHttpXForwardedFor(forwardedHops);
+            // If multiple IPs available, the first ip is the client
+            int firstCommaIndex = forwardedHops.indexOf(',');
+            outboundAccessLogMessage.setIp(firstCommaIndex != -1 ?
+                    forwardedHops.substring(0, firstCommaIndex) : forwardedHops);
+        }
+        if (headers.contains(HttpHeaderNames.USER_AGENT)) {
+            outboundAccessLogMessage.setHttpUserAgent(headers.get(HttpHeaderNames.USER_AGENT));
+        }
+        if (headers.contains(HttpHeaderNames.REFERER)) {
+            outboundAccessLogMessage.setHttpReferrer(headers.get(HttpHeaderNames.REFERER));
+        }
+        HttpAccessLogConfig.getInstance().getCustomHeaders().forEach(customHeader ->
+                outboundAccessLogMessage.putCustomHeader(customHeader, headers.contains(customHeader) ?
+                        headers.get(customHeader) : "-"));
+
+        outboundAccessLogMessage.setRequestMethod(httpOutboundRequest.getHttpMethod());
+        outboundAccessLogMessage.setRequestUri((String) httpOutboundRequest.getProperty(TO));
+        HttpMessage inboundResponse = inboundResponseMsg.getNettyHttpResponse();
+        if (inboundResponse != null) {
+            outboundAccessLogMessage.setScheme(inboundResponse.protocolVersion().toString());
+        } else {
+            outboundAccessLogMessage.setScheme(inboundResponseMsg.getHttpVersion());
+        }
+        outboundAccessLogMessage.setRequestBodySize((long) httpOutboundRequest.getContentSize());
+        outboundAccessLogMessage.setStatus(inboundResponseMsg.getHttpStatusCode());
+        outboundAccessLogMessage.setResponseBodySize(contentLength);
+        long requestTime = Calendar.getInstance().getTimeInMillis() -
+                outboundAccessLogMessage.getDateTime().getTimeInMillis();
+        outboundAccessLogMessage.setRequestTime(requestTime);
+
+        HttpCarbonMessage inboundReqMsg =
+                getTypedProperty(httpOutboundRequest, INBOUND_MESSAGE, HttpCarbonMessage.class);
+
+        if (inboundReqMsg != null) {
+            List<HttpAccessLogMessage> outboundAccessLogMessages = getHttpAccessLogMessages(inboundReqMsg);
+            if (outboundAccessLogMessages != null) {
+                outboundAccessLogMessages.add(outboundAccessLogMessage);
+            }
+        }
+    }
+
+    private <T> T getTypedProperty(HttpCarbonMessage request, String propertyName, Class<T> type) {
+        Object property = request.getProperty(propertyName);
+        if (type.isInstance(property)) {
+            return type.cast(property);
+        }
+        return null;
+    }
+
+    private List<HttpAccessLogMessage> getHttpAccessLogMessages(HttpCarbonMessage request) {
+        Object outboundAccessLogMessagesObject = request.getProperty(OUTBOUND_ACCESS_LOG_MESSAGES);
+        if (outboundAccessLogMessagesObject instanceof List<?> rawList) {
```

Comment:
looks like we are doing the same thing in multiple places. let's remove if its redundant as I said in the above comment or else lets move it to a the State parent class.